File: 2 курс/Алгоритмы и структуры данных/tasksPython/G.py
s = input()
freq = [0] * 26
sorted = []
weights = list(map(int, input().split()))

print(weights)
lettersWithWeights = {}


def getChar(index):
    return chr(ord('a') + index)


for i in range(len(s)):
    sorted.append(s[i])
for _ in range(len(sorted) - 1):
    for i in range(len(sorted) - 1):
        if sorted[i] < sorted[i + 1]:
            fills = sorted[i + 1]
            sorted[i + 1] = sorted[i]
            sorted[i] = fills

ans_i = len(s)
ans = [""] * ans_i

for n in sorted:
    freq[ord(n) - 97] += 1

print(freq)
fills = []
candidates = []
for i in range(26):
    char = getChar(i)
    fr = freq[i]
    if fr >= 2:
        candidates.append((weights[i], char, fr))
    else:
        fills.append(char)

candidates.sort(reverse=True)
print(candidates)

n = len(ans)
left = 0
right = n - 1
for weight, char, freq in candidates:
    ans[left] = char
    ans[right] = char
    left += 1
    right -= 1
    for _ in range(freq - 2):
        fills.append(char)

j = 0
for i in range(len(ans)):
    if ans[i] == "":
        ans[i] = fills[j]
        j += 1

print("".join(ans))

File: 2 курс/Алгоритмы и структуры данных/tasksCpp/G.cpp
// #include <algorithm>
// #include <iostream>
// #include <numeric>
// #include <string>
// #include <vector>
//
// using namespace std;
//
// int main() {
//   string s;
//   cin >> s;
//
//   vector<long long> weight(26);
//   for (int i = 0; i < 26; ++i) {
//     cin >> weight[i];
//   }
//
//   vector<int> freq(26, 0);
//   for (char c : s) {
//     freq[c - 'a']++;
//   }
//
//   vector<int> order(26);
//   iota(order.begin(), order.end(), 0);
//   sort(order.begin(), order.end(), [&](int a, int b) {
//     if (weight[a] != weight[b])
//       return weight[a] > weight[b];
//     return a < b;
//   });
//
//   int n = (int)s.size();
//   string ans(n, '\0');
//   int left = 0, right = n - 1;
//
//   while (true) {
//     vector<int> candidates;
//     for (int id : order) {
//       if (freq[id] >= 2) {
//         candidates.push_back(id);
//       }
//     }
//
//     if (candidates.empty() || left > right) {
//       break;
//     }
//
//     for (int id : candidates) {
//       if (left > right)
//         break;
//
//       ans[left++] = char('a' + id);
//       ans[right--] = char('a' + id);
//       freq[id] -= 2;
//     }
//   }
//
//   string middle;
//   for (int id : order) {
//     while (freq[id] > 0) {
//       middle.push_back(char('a' + id));
//       --freq[id];
//     }
//   }
//
//   int idx = 0;
//   for (int i = 0; i < n; ++i) {
//     if (ans[i] == '\0') {
//       ans[i] = middle[idx++];
//     }
//   }
//
//   cout << ans << '\n';
//   return 0;
// }